### PR TITLE
feat: eip-6963 support

### PR DESCRIPTION
**Pull request by gomesalexandre**
a.k.a "The Big Short (2015)".

⚠️ Ready for conceptual review and early testing with its web companion, but **not** ready to merge until web is ready - few things to tackle there both from a technical and product perspective.
Most likely, nothing will change here in hdwallet-land as the changeset is extremely low and we shouldn't require anything more, though some smolish follow-ups could be made:

- Phantom leverages EIP provider
- Keplr EVM support is added through EIP provider

This PR brings EIP-6963 support through passing an rdns when instantiating metamask, and detecting the correct provider instead of `detectEthereumProvider`. That's it.
All the other changes here are effectively:

- removal of unused hdwallet-metamask (flag was off in web for a long time already and we weren't hitting its branch)
- removal of now unused hdwallet-xdefi
- removal of now unused hdwallet-coinbase
- removal of unused hdwallet-tallyho

See inline comments for the only actual change in this mist of red lines.

### CODE REVIEW

File: examples/sandbox/index.html
Comment (gomesalexandre) [inline examples/sandbox/index.html:123]: Removed across the board in this PR:

- Tally was dead for a long time already, and could ironically revive if Tally announced themselves as en EIP-6363 provider https://github.com/tahowallet/extension/issues/3725
- Coinbase doesn't need an explicit button anymore and is handled by the EIP detection
- Same as above for XDEFI

image

File: examples/sandbox/index.ts
Comment (gomesalexandre) [inline examples/sandbox/index.ts:12]: We weren't relying on hdwallet-metamask in web for a long, very long time already. Ditches the regular hdwallet-metamask since the multichain one superseeds it.

File: packages/hdwallet-metamask-shapeshift-multichain/src/adapter.ts
Comment (gomesalexandre) [inline packages/hdwallet-metamask-shapeshift-multichain/src/adapter.ts:1]: We obviously don't need this anymore since we support MIPD

File: packages/hdwallet-metamask-shapeshift-multichain/src/adapter.ts
Comment (gomesalexandre) [inline packages/hdwallet-metamask-shapeshift-multichain/src/adapter.ts:94]: Yep, that's all it takes. Didn't even need to pass it from web in the end (which didn't work, something something proxies?), which is actually good, in-place provider detection with rdns

Comment (gomesalexandre) [reply packages/hdwallet-metamask-shapeshift-multichain/src/adapter.ts:94]: Note naming: EIP-6963 structure here, with `provider` (an EIP-1193 compliant provider) and `info` (provider info) vs. 1193 below, the good ol' JS provider API https://eips.ethereum.org/EIPS/eip-1193

File: packages/hdwallet-metamask-shapeshift-multichain/package.json
Comment (gomesalexandre) [inline packages/hdwallet-metamask-shapeshift-multichain/package.json:25]: For ease of use instead of dealing with EIP-6963 reqs/event handlers ourselves, great lib from wevm as usual!

File: packages/hdwallet-trezor/src/trezor.ts
Comment (kaladinlight) [inline packages/hdwallet-trezor/src/trezor.ts:153]: ? part of `BTCWalletInfo` interface. seems like we should still have this in some form or fashion

File: packages/hdwallet-core/src/bitcoin.ts
Comment (kaladinlight) [inline packages/hdwallet-core/src/bitcoin.ts:348]: oh I see you removed it. major version bump change to interface? (no idea what the point of this function was for)

Comment (gomesalexandre) [reply packages/hdwallet-core/src/bitcoin.ts:348]: Yeah given the fact that we have removed quite a big of packages, we should definitely go with at the very least a minor vs. a patch as we usually do. Not sure about major here, something something about open-source. 

Looking at Trezor implementation, this is used to detect diff. purposes over same slip44 + accountId... which we can do, and actually do pretty easily in web already! 
image

File: packages/hdwallet-native/src/bitcoin.ts
Comment (kaladinlight) [inline packages/hdwallet-native/src/bitcoin.ts:38]: dont' forget

File: packages/hdwallet-metamask-multichain/src/adapter.ts
Comment (kaladinlight) [inline packages/hdwallet-metamask-multichain/src/adapter.ts:6]: we should update class name to match while we are at it

**Comment by kaladinlight** [comment]
Main concern coming to mind with this is the fact that the other eip-6963 providers will now be "multichain", when in reality they aren't. all of our logic is dependent on the `supportsCoin` logic which wouldn't be accurate. This was why I was thinking a base eip-6963 implementation with the injected provider, and then everyone can extend that as they see fit (multichain, phantom, keplr, etc). I suppose that isn't entirely necessary for the phantoms/keplers since we will be instantiating their specific wallet in web, but for xdefi/coinbase, it seems applicable in that we would be using a wallet with multichain support when it doesn't have it (unless I am missing something).

**Comment by gomesalexandre** [comment]
> Main concern coming to mind with this is the fact that the other eip-6963 providers will now be "multichain", when in reality they aren't. all of our logic is dependent on the `supportsCoin` logic which wouldn't be accurate. This was why I was thinking a base eip-6963 implementation with the injected (or detected) provider, and then everyone can extend that as they see fit (multichain, phantom, keplr, etc). I suppose that isn't entirely necessary for the phantoms/keplers since we will be instantiating their specific wallet in web, but for xdefi/coinbase, it seems applicable in that we would be using a wallet with multichain support when it doesn't have it (unless I am missing something).

As discussed with @kaladinlight, hdwallet improvements could be done as a follow-up (though hdwallet is perma-improvements land), but as it stands now, there is no change re: current flow, since we weren't using the regular `hdwallet-metamask` anymore for a long, long time already. See description for two follow-ups that would make sense effort-wise, but probably not worth doing any kind of MM snaps/pure EVM hdwallet split in hdwallet land as web is handling this at runtime anyway.

File: packages/hdwallet-native/src/bitcoin.ts
Comment (gomesalexandre) [reply packages/hdwallet-native/src/bitcoin.ts]: 27297629e92051d6442f056b0da56dd8a915a0f2 - keeping opened in case it comes back biting us to nuke again

File: packages/hdwallet-metamask-multichain/src/adapter.ts
Comment (gomesalexandre) [reply packages/hdwallet-metamask-multichain/src/adapter.ts]: a9fa9f5079c02d12f9800e9e182b88a5f8ffa141 (don't mind the diff, needed actual bimp for `integration` not to be borked... which I also added missing package.json entry for [there](https://github.com/shapeshift/hdwallet/commit/a9fa9f5079c02d12f9800e9e182b88a5f8ffa141#diff-cd67598767aad4b2e7c64086173ce0c83943deb401b2a81f8e3d835d8141cbc9R18), was shadowed previously)

File: examples/sandbox/package.json
Comment (gomesalexandre) [inline examples/sandbox/package.json:24]: Sandbox is a big mess - as far as I can understand from these stale refs, this doesn't use lerna but upstream (where `@shapeshiftoss/hdwallet-metamask-multichain` only exists as `1.55.10-mipd.0` alpha for now) , and trying to use 1.55.10 across the board here is a rug:
image

So looks like we're always one version behind in this package.json pretty much?

Comment (gomesalexandre) [reply examples/sandbox/package.json]: Actually shadowed *is* fine, that's what we want here not to always be one version behind and have to manually update: 7309f547
image

File: packages/hdwallet-metamask-multichain/src/adapter.ts
Comment (gomesalexandre) [reply packages/hdwallet-metamask-multichain/src/adapter.ts]: Actually: https://github.com/shapeshift/hdwallet/pull/689#discussion_r1787469657

File: examples/sandbox/package.json
Comment (gomesalexandre) [reply examples/sandbox/package.json]: Actually, shadowed *is* fine and that's the reason why things were borked in the first place with this package rename, and why we were always a version behind and had to manually update. 7309f547e9efc247c128c560c64ea1fe19dcf1a4

File: packages/hdwallet-core/package.json
Comment (kaladinlight) [inline packages/hdwallet-core/package.json:30]: remove the `gitHead`s from all of the package jsons

File: examples/sandbox/package.json
Comment (kaladinlight) [inline examples/sandbox/package.json:21]: sandbox should ref 1.55.11 too

File: integration/src/wallets/metamask.ts
Comment (kaladinlight) [inline integration/src/wallets/metamask.ts:63]:
```suggestion
  it("supports Bitcoin", async () => {
```

File: integration/src/wallets/metamask.ts
Comment (kaladinlight) [inline integration/src/wallets/metamask.ts:73]:
```suggestion
  it("does not support bip44 accounts", async () => {
```

File: integration/src/wallets/metamask.ts
Comment (kaladinlight) [inline integration/src/wallets/metamask.ts:92]: revert. no multi account wallet support, but can still get and describe paths as it stands.

File: integration/src/wallets/metamask.ts
Comment (kaladinlight) [inline integration/src/wallets/metamask.ts:95]: revert, pref to use `coin` as it appears in the `slip44Table`. can `toLowerCase` the `describePath` switch statement in `shapeshift-multichain` or just update to coin keys (uppercase)

File: packages/hdwallet-metamask-multichain/src/adapter.ts
Comment (kaladinlight) [inline packages/hdwallet-metamask-multichain/src/adapter.ts:27]: we lost the onboarding logic here, do we want to trigger mm onboarding for `io.metamask` so things behave the same?

File: packages/hdwallet-metamask-multichain/src/mipdStore.ts
Comment (kaladinlight) [inline packages/hdwallet-metamask-multichain/src/mipdStore.ts:3]: save a file and just instantiate in the adapter?

File: packages/hdwallet-metamask-multichain/src/adapter.ts
Comment (gomesalexandre) [reply packages/hdwallet-metamask-multichain/src/adapter.ts:27]: Waiting on a reply from product here, intent is to use wc instead since deep linking is broken half/most of the time depending on platforms

File: packages/hdwallet-core/package.json
Comment (gomesalexandre) [reply packages/hdwallet-core/package.json]: e8d0d107

File: integration/src/wallets/metamask.ts
Comment (gomesalexandre) [reply integration/src/wallets/metamask.ts]: e8d0d107

File: integration/src/wallets/metamask.ts
Comment (gomesalexandre) [reply integration/src/wallets/metamask.ts]: e8d0d107

File: packages/hdwallet-metamask-multichain/src/mipdStore.ts
Comment (gomesalexandre) [reply packages/hdwallet-metamask-multichain/src/mipdStore.ts]: e8d0d107

File: examples/sandbox/package.json
Comment (gomesalexandre) [reply examples/sandbox/package.json:21]: image

Doesn't look like we're able to do it in the same PR and are inherently one version ahead?

Comment (kaladinlight) [reply examples/sandbox/package.json:21]: we removed coinbase so that is why it isn't able to ref a local package

File: packages/hdwallet-metamask-multichain/src/adapter.ts
Comment (gomesalexandre) [reply packages/hdwallet-metamask-multichain/src/adapter.ts:29]: Disregard me - confirmed we want to keep the current behaviour *fully* as in, always display providers even if not installed (in the same order as currently), only appending the ones we don't currently have already

File: packages/hdwallet-metamask-shapeshift-multichain/src/adapter.ts
Comment (gomesalexandre) [reply packages/hdwallet-metamask-shapeshift-multichain/src/adapter.ts:1]: Actually we do

File: integration/src/wallets/metamask.ts
Comment (gomesalexandre) [reply integration/src/wallets/metamask.ts]: df02ebb

File: integration/src/wallets/metamask.ts
Comment (gomesalexandre) [reply integration/src/wallets/metamask.ts]: df02ebb2

File: examples/sandbox/package.json
Comment (gomesalexandre) [reply examples/sandbox/package.json]: 🧠 de2247e6

File: packages/hdwallet-metamask-multichain/src/shapeshift-multichain.ts
Comment (kaladinlight) [inline packages/hdwallet-metamask-multichain/src/shapeshift-multichain.ts:97]: really feels like we should do this to all of the coins in here to make sure they match the `slip44Table` (not that we really use this function anywhere, but for posterity)

File: integration/src/wallets/metamask.ts
Comment (kaladinlight) [inline integration/src/wallets/metamask.ts:91]: this intergration test is only for metamask anyways, so ternary is kind of pointless. can just have `[0]` if we don't want to show that it still _is_ able to derive account paths even if not natively supported by the wallet itself

File: packages/hdwallet-metamask-multichain/src/adapter.ts
Comment (kaladinlight) [inline packages/hdwallet-metamask-multichain/src/adapter.ts:30]: we don't want to trigger the metamask onboarding if trying to pair a non MM wallet (xdefi/coinbase/tally/etc). otherwise just generic throw.
`if (!maybeEip6963Provider && isMetaMask) {`

File: packages/hdwallet-metamask-shapeshift-multichain/src/adapter.ts
Comment (kaladinlight) [reply packages/hdwallet-metamask-shapeshift-multichain/src/adapter.ts:1]: ?

File: packages/hdwallet-metamask-multichain/src/shapeshift-multichain.ts
Comment (gomesalexandre) [reply packages/hdwallet-metamask-multichain/src/shapeshift-multichain.ts:97]: All in for cleanup! e0e8f4e587c7b8143c9603066ad92ab08fe08762

File: integration/src/wallets/metamask.ts
Comment (gomesalexandre) [reply integration/src/wallets/metamask.ts]: e0e8f4e5

File: packages/hdwallet-metamask-multichain/src/adapter.ts
Comment (gomesalexandre) [reply packages/hdwallet-metamask-multichain/src/adapter.ts]: e0e8f4e5

Comment (kaladinlight) [reply packages/hdwallet-metamask-multichain/src/adapter.ts]: We want to still throw a generic error on `!maybeEip6963Provider`?

Comment (gomesalexandre) [reply packages/hdwallet-metamask-multichain/src/adapter.ts]: Absolutely - as discussed this wasn't really a concern previously (if we ever hit this, we could safely assume there *is* a provider since its rdns was originally detected in web) but since we're now adding a static list of mipd providers (regardless of installation), then we absolutely want to throw here in case we hit a static non-MetaMask rdns (Coinbase, XDEFI, Rabby) and it isn't installed: 1b4ad64b

File: examples/sandbox/package.json
Comment (kaladinlight) [inline examples/sandbox/package.json:23]: don't forget to change before merge

**Review by kaladinlight — APPROVED** [review]

**Pull request merged**
